File: HW7 Java/hw7_1901042694/src/DayofYearSet.java
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
public class DayofYearSet{
    /*
       *Inner class
     */
    public static class DayofYear{
        private int day;
        private int month;
        public DayofYear(int Day, int Month) {
            this.day = Day;
            this.month = Month;
            ++alive;
        }
        public DayofYear() {}
        public int getDay() {
            return this.day;
        }
        public int getMonth() {
            return this.month;
        }
        public void setAlive(int Alive){alive=Alive;}
        public int getAlive(){ return alive;}
        public void setDay(int day) {
            this.day = day;
        }
        public void setMonth(int month) {
            this.month = month;
        }
        @Override
        public String toString() {
            return "DayofYear{" + "day=" + day + ", month=" + month + '}';
        }

    }
    /*
     *Constructors;
     */
    public DayofYearSet() {}
    public DayofYearSet(ArrayList<DayofYear> data) {
        //Casting our parameter as a array
            list= data.toArray(new DayofYear[0]);
            used=data.size();

    }
    /*
    *Getter and Setter
     */
    public DayofYear[] getList() {
        return this.list;
    }
    public void setList(DayofYear list[] ) {
        this.list = list;
    }
    public void setData(DayofYearSet.DayofYear data) {
        this.data = data;
    }
    public DayofYearSet.DayofYear getData() {
        return this.data;
    }
    public int getSize(){return used;}
    public void setSize(int newSize){used=newSize;}
    /*
     *Main Methods
     */
    public void delete(DayofYear item) {
        int index=findIndex(item);
        int size=used;
        if(checkInput(item.getDay(),item.getMonth())){
            DayofYear[] newList = new DayofYear[used];
            if(index!=-1){
                int i=0,j=0;
                while(i<used){
                    if(i!=index){
                        newList[j]=list[i];
                        ++j;
                    }
                    ++i;
                }
                System.out.println(j+" Size");
                setList(newList);
                setSize(j);
            }
        }
    }

    /*
     *To find difference between two objects
     */
    public DayofYearSet setDifference(DayofYearSet other){
        DayofYearSet newList=new DayofYearSet();
        var size=list.length;
        int i;
        for(i=0;i<size;++i){
            if(!duplicate(other.list,list[i])){
                newList.add(list[i]);
            }
        }
        return newList;
    }
    /*
    *Add an item to DayofYearSet Object
    *First off all , We check input and used if used is less than or equal 1 we do not have to check whether the item is duplicated or not.
     */
    public void add(DayofYear item) {
        if(checkInput(item.getDay(),item.getMonth())){
            if(used!=0){
                if (!this.duplicate(this.list,item)) {
                    this.data.day = item.day;
                    this.data.month = item.month;
                    this.list[used]=item;
                    ++used;

                }
            }else{

                this.data.day = item.day;
                this.data.month = item.month;
                this.list[used]=item;
                ++used;
            }
        }
    }
    /*
        *Static function that counts all alive DayofYear object;
     */
    public static int totalAlive(){return alive;}
    /*
    *This function finds whether these 2 objects are equal or not without looking order;
     */
    public boolean equals(DayofYearSet other) {
        var size=this.used;
        var size2=other.used;
        int i,j;
        int count=0;
        var flag=false;
        if(size==size2){
            for(i=0;i<size;++i){
                for(j=0;j<size2;++j){
                    if(equalData(list[i],other.list[j])){
                        flag=true;
                        ++count;
                    }
                }
            }
            if(count==size){
                flag=true;
            }else{
                flag=false;
            }
        }
        return flag;
    }

    public DayofYearSet setComplement(){
        DayofYearSet cagri=new DayofYearSet();
        DayofYear[] returnedValue=setDayOfYearList();
        int i,j=0;
        boolean flag=false;
        int index=0;
        for(j=0;j<365;++j){
            for(i=0;i<used;++i){
                if(equalData(list[i],returnedValue[j])){
                    flag=true;
                    i=used;
                }
            }
            if(flag){
                ++index;
                flag=false;
            }else{
                cagri.add(returnedValue[j]);
            }
        }
        cagri.setSize(365-index);
        cagri.setList(cagri.list);
        return cagri;

    }
    /*
    *Helper Functions
    * This function controls that two DayofYearSet objects are equal or not
    * This function prevents an item adds at least two times
     */
    public boolean duplicate(DayofYear List[] , DayofYear element) {
        int i ;
        boolean flag = false;

        for(i = 0; i < getSize(); ++i) {
            if (this.equalData(List[i], element)) {
                flag = true;
                break;
            }
            flag = false;

        }
        return flag;
    }
    /*
    *This function prevents to enter wrong input;
     */
    public boolean checkInput(int day,int month){
        boolean flag=false;
        if(month==1){
            if(day>=1&&day<=31){
                flag=true;
            }
        }
        if(month==2){
            if(day>=1&&day<=28){
                flag=true;
            }
        }
        if(month==3){
            if(day>=1&&day<=31){
                flag=true;
            }
        }

        if(month==4){
            if(day>=1&&day<=30){
                flag=true;
            }
        }
        if(month==5){
            if(day>=1&&day<=31){
                flag=true;
            }
        }
        if(month==6){
            if(day>=1&&day<=30){
                flag=true;
            }
        }
        if(month==7){
            if(day>=1&&day<=31){
                flag=true;
            }
        }
        if(month==8){
            if(day>=1&&day<=31){
                flag=true;
            }
        }
        if(month==9){
            if(day>=1&&day<=30){
                flag=true;
            }
        }
        if(month==10){
            if(day>=1&&day<=31){
                flag=true;
            }
        }
        if(month==11){
            if(day>=1&&day<=30){
                flag=true;
            }
        }
        if(month==12){
            if(day>=1&&day<=31){
                flag=true;
            }
        }
        return flag;
    }
    /*
     *This function controls two DayOfyear items are equal or not
     */
    public boolean equalData(DayofYearSet.DayofYear item, DayofYearSet.DayofYear item1) {
        return item1.day == item.day && item1.month == item.month;
    }
    /*
     *These function eases to find index that wants to remove an item from list;
     */
    public int findIndex(DayofYear item){
        int index=-1;
        for(int i = 0; i < this.list.length; ++i) {
            if (this.equalData((DayofYearSet.DayofYear)this.list[i], item)) {
                index=i;
                break;
            }
        }
        return index;
    }
    /*
     *This function adds a DayofYearSet object to the other object
     */
    public DayofYearSet setUnion(DayofYearSet other) {
        int size2=other.getSize();
        System.out.println(size2);
        int j;
        for(j=0;j<size2;++j){
            if(!duplicate(this.list,other.list[j])){
                add(other.list[j]);

            }
        }
        return this;
    }
    /*
     *To find the common item
     */
    public DayofYearSet setIntersection(DayofYearSet other){
        DayofYearSet returnedValue=new DayofYearSet();
        int size=other.used;

        if(used> other.used){
            size=used;
        }
        int i;
        for(i=0;i< other.used;++i){
            if(duplicate(this.list,other.getList()[i])){
                returnedValue.add(other.getList()[i]);
            }
        }
        return returnedValue;
    }
    /*
     *This function needs to complement list because we need to make an array that holds all days of year it returns a list that has all days of year;
     */
    public DayofYear[] setDayOfYearList(){
        DayofYear[] list = new DayofYear[365] ;
        int i=1,j=1;
        int index=0;
        for(i=1;i<=12;++i){
            if(i==1||i==3||i==5||i==7||i==8||i==10||i==12){
                for(j=1;j<=31;++j){
                    DayofYear data=new DayofYear(j,i);
                    list[index]=data;
                    ++index;
                }
            }
            if(i==4||i==6||i==9||i==11){
                for(j=1;j<=30;++j){
                    DayofYear data=new DayofYear(j,i);
                    list[index]=data;
                    ++index;
                }
            }
            if(i==2){
                for(j=1;j<=28;++j){
                    DayofYear data=new DayofYear(j,i);
                    list[index]=data;
                    ++index;
                }
            }
            j=1;
        }
        return list;
    }
    void writeToFile(String fileName){
        try {
            File myObj = new File(fileName);
            if (myObj.createNewFile()) {
                System.out.println("File created: " + myObj.getName()+"\n");
            } else {

                System.out.println("File already exists.");
            }
            try {
                int i;
                FileWriter writer = new FileWriter(fileName);
                writer.write("Day\t"+"Month\n");
                for(i=0;i<used;++i){
                    writer.write(" "+list[i].getDay()+"\t"+list[i].getMonth()+" \n");
                }
                writer.write("Active Games "+alive);
                writer.close();
                System.out.println("Successfully wrote to the file.");
            } catch (IOException e) {
                System.out.println("Error");
                e.printStackTrace();
            }
        } catch (IOException e) {
            System.out.println("An error occurred.");
            e.printStackTrace();
        }
    }
    /*
       *To write details about DayofYearSet Objects;
     */
    @Override
    public String toString() {
        StringBuilder returnedValue= new StringBuilder();
        returnedValue.append("Size "+used+"\n");
        int i;
        for(i=0;i<used;++i){
            returnedValue.append("Day: "+list[i].day+" Month "+list[i].month+" \n");
        }
        return returnedValue.toString();
    }
    /*
     *Field
     */
    private DayofYear[] list = new DayofYear[365] ;
    private DayofYearSet.DayofYear data = new DayofYearSet.DayofYear();
    private int size=0;
    /*
     *How many item that we put into the object;
     */
    private int used=0;
    /*
     *To find that how many DayofYear objects are alive
     */
    private static int alive=0;
}

File: HW4/DayOfYearSet.cpp
#include "DayOfYearSet.h"

namespace cagri{


	int  DayofYearSet::activeGames = 0;
	int  DayofYearSet::NumberofActiveGames() { return activeGames; }
	 DayofYearSet::DayofYear&  DayofYearSet:: operator[](int index) {
		 DayofYearSet::DayofYear* list = getDatas();
		if (index < 0) {
			 std::cerr << "Wrong index";
			exit(1);
		}
		else {
			return list[index];
		}
	}
	bool   operator !=( DayofYearSet datasMember1,  DayofYearSet datasMember2) {
		bool flag = false;
		if (!datasMember1.areTheSameSets(datasMember2)) {
			flag = true;
		}

		return flag;
	}
	bool   operator ==( DayofYearSet  datasMember1, DayofYearSet  datasMember2) {
		bool flag = false;
		if (datasMember1.getSize() != datasMember2.getSize()) {
			flag = false;
		}
		else {
			if (datasMember1.areTheSameSets(datasMember2)) {
				flag = true;
			}
		}
		return flag;
	}
	 DayofYearSet  operator +( DayofYearSet datasMember1,  DayofYearSet datasMember2) {
		 DayofYearSet list;
		// DayofYearSet::DayofYear* newList = list.getDatas();
		auto size1 = datasMember1.getSize();
		decltype(size1) size2 = datasMember2.getSize();
		int i;
		for (i = 0;i < size2;++i) {
			datasMember1.add(datasMember2.getDatas()[i]);
		}
		datasMember1.setSize(size1 + size2);
		list = datasMember1;
		return list;

	}
	void  DayofYearSet::setList(DayofYear* datasMember1) {
		datasMember = datasMember1;
	}

	 DayofYearSet::DayofYearSet(vector<DayofYear> data) {
		++activeGames;
		datasMember = new DayofYear[365];

	}
	 DayofYearSet::DayofYear::DayofYear(int Month, int Day) : month(Month), day(Day) {

	}
	 DayofYearSet::DayofYear::DayofYear() {

	}
	 DayofYearSet  operator -( DayofYearSet newDataMembers,  DayofYearSet newDataMembers1) {
		int i, j;
		int firstSize = newDataMembers.getSize();
		int secondSize = newDataMembers1.getSize();
		bool flag = false;
		 DayofYearSet returnedList;

		for (i = 0;i < firstSize;++i) {
			for (j = 0;j < secondSize;++j) {
				if (newDataMembers.isEqual(newDataMembers.getDatas()[i], newDataMembers1.getDatas()[j])) {
					flag = true;
					j = secondSize;
				}

			}
			if (flag == false) {
				returnedList.add(newDataMembers.getDatas()[i]);
			}
		}
		return returnedList;
	}
	bool  DayofYearSet::isEqual(DayofYearSet::DayofYear first, DayofYearSet::DayofYear second) {
		return first.getDay() == second.getDay() && first.getMonth() == second.getMonth();
	}
	bool  DayofYearSet::duplicate(DayofYear element) {
		int i;
		bool flag = false;
		DayofYear* data = getDatas();
		int size1 = getSize();

		if (size1 > 0) {
			for (i = 0;i < size1;++i) {

				if (isEqual(data[i], element)) {
					flag = true;
					break;
				}
				else {}
			}
		}
		else if (size1 == 0) {
			flag = false;
		}
		return flag;
	}
	bool  DayofYearSet::areTheSameSets(DayofYearSet second) {
		DayofYear* firstObj = getDatas();
		DayofYear* secondObj = second.getDatas();
		bool flag = false;
		int size1 = getSize();
		if (size1 > 0 && second.getSize() > 0) {
			int i, j;
			for (i = 0;i < size1;++i) {
				for (j = 0;j < second.getSize();++j) {
					if (isEqual(firstObj[i], secondObj[j])) {
						flag = true;
						j = second.getSize();
					}
					else {
						flag = false;
					}

				}
			}
		}
		return flag;
	}


	 DayofYearSet::DayofYearSet() {
		++activeGames;
		datasMember = new DayofYear[1000];

	}
	 DayofYearSet  operator ^( DayofYearSet newDataMembers,  DayofYearSet newDataMembers1) {
		int i, j;
		int firstSize = newDataMembers.getSize();
		int secondSize = newDataMembers1.getSize();

		bool flag = false;
		//int a = 0;
		//const int size = newDataMembers.getSize() + newDataMembers1.getSize();
		 DayofYearSet returnedList;


		for (i = 0;i < firstSize;++i) {

			for (j = 0;j < secondSize;++j) {
				if (newDataMembers.isEqual(newDataMembers.getDatas()[i], newDataMembers1.getDatas()[j])) {
					flag = true;


					j = secondSize;

				}
				else { flag = false; }

			}
			if (flag == true) {
				returnedList.add(newDataMembers.getDatas()[i]);
			}
		}
		return returnedList;
	}

	 DayofYearSet  operator !( DayofYearSet dataMember1) {
		int  j, z, k = 0;
		 DayofYearSet::DayofYear* list = dataMember1.getDatas();
		int size = dataMember1.getSize();

		//to remark day of a year;
		int dayOfaYear[][31] = {
			{1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31}
			,{1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,-1,-1,-1}
			,{1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31}
			,{1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,-1}
			,{1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31}
			,{1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,-1}
			,{1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31}
			,{1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31}
			,{1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,-1}
			,{1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31}
			,{1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,-1}
			,{1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31} };
		 DayofYearSet newDatas;
		
		 DayofYearSet::DayofYear* newList = new  DayofYearSet::DayofYear[365];

		// This loop finds that complement of our object
		int q = 0;

		bool flag = false;
		for (j = 0;j < 12;++j) {
			for (z = 0;z < 31;++z) {
				if (dayOfaYear[j][z] != -1) {
					for (q = 0;q < size;++q) {
						int day = list[q].getDay();
						--day;
						int month = list[q].getMonth();
						--month;
						if (day == z && month == j) {
							flag = false;
							q = size;
						}
						else {
							flag = true;

						}
					}
					if (flag == true) {

						newList[k].setDay(z);
						newList[k].setMonth(j);
						++k;
						flag = false;
					}
				}
			}
		}
		newDatas.setList(newList);
		newDatas.setSize(k);
		return newDatas;
	}
	void  DayofYearSet::add(DayofYear element) {
		DayofYear* datasMember = getDatas();
		int size1 = getSize();
		if (size1 < 1000) {
			if (!duplicate(element)) {
				*(datasMember + size1) = element;
				setList(datasMember);
				setSize(size1 + 1);
			}

		}
		else {
			 std::cout << "Buraya girdi\n";
		}

	}

	void  DayofYearSet::remove(const int indeks) {
		DayofYear* list = getDatas();
		int i;
		int j = 0;
		const int size = getSize();
		DayofYear* list1 = new DayofYear[size];
		if (indeks < size) {
			if (size > 1) {
				for (i = 0;i < size;++i) {
					if (i < indeks) {
						*(list1 + j) = *(list + i);
						++j;
					}
					else {
						*(list1 + j) = *(list + i + 1);
						j++;
					}
				}
			}
			else {}
			setList(list1);
			setSize(j - 1);
		}

	}
}
